### PR TITLE
Contribute to the RMT Chapter 

### PR DESCRIPTION
**Pull request by farrosalferro**
The following is the content of the RMT.mdx:

- What is retention networks
   Overview of what is the motivation behind retention networks, the mechanism and how does it resolve the LLM's challenges.
- From Language to Image
   Introduce RMT and ViR, two computer vision models that are inspired by retention networks.
- Further Reading
   Extra resources to gain better understanding about retention networks, RMT, ViR.

### CODE REVIEW

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Retention In Vision
+
+## What is Retention Networks
+RetNet, or Retentive Network, is a foundational architecture proposed for large language models in the paper [Retentive Network: A Successor to Transformer for Large Language Models](https://arxiv.org/abs/2307.08621). This architecture is designed to address key challenges in the realm of large-scale language modeling: Training Parallelism, Low-Cost Inference, and Good performance.
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:4]:
```suggestion
Retentive Network (RetNet) is a foundational architecture proposed for large language models in the paper [Retentive Network: A Successor to Transformer for Large Language Models](https://arxiv.org/abs/2307.08621). This architecture is designed to address key challenges in the realm of large-scale language modeling: training parallelism, low-cost inference, and good performance.
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+![LLM Challenges.png](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/LLM%20Challenges.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:6]:
```suggestion
![LLM Challenges](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/LLM%20Challenges.png)
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+RetNet is able to tackle these challenges by introducing the Multi-Scale Retention (MSR) mechanism, which is an alternative to the multi-head attention mechanism commonly used in Transformer models. 
+MSR has a dual form of recurrence and parallelism, so it is possible to train the models in a parallel way while recurrently conducting inference.
+
+The Multi-Scale Retention mechanism operates under three computation paradigms:
+1) **Parallel Representation**: This aspect of RetNet is designed to fully utilize GPU resources, enabling efficient parallel training of the model.
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:11]:
```suggestion
- **Parallel Representation:** This aspect of RetNet is designed to fully utilize GPU resources, enabling efficient parallel training of the model.
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+![Parallel Representation.png](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/Parallel%20Representation.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:12]:
```suggestion
![Parallel Representation](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/Parallel%20Representation.png)
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+2) **Recurrent Representation**: This representation facilitates efficient inference with O(1) complexity in terms of memory and computational requirements. It significantly reduces deployment costs and latency, and simplifies implementation by eliminating the need for key-value cache strategies often used in traditional models.
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:14]:
```suggestion
- **Recurrent Representation:** This representation facilitates efficient inference with O(1) complexity in terms of memory and computational requirements. It significantly reduces deployment costs and latency, and simplifies implementation by eliminating the need for key-value cache strategies often used in traditional models.
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+![Recurrent Representation.png](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/Recurrent%20Representation.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:15]:
```suggestion
![Recurrent Representation](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/Recurrent%20Representation.png)
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+3) **Chunkwise Recurrent Representation**: This third paradigm addresses the challenge of long-sequence modeling. It achieves this by encoding each local block in parallel for computational speed while simultaneously encoding global blocks in a recurrent manner to optimize GPU memory usage.
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:17]:
```suggestion
- **Chunkwise Recurrent Representation:** This third paradigm addresses the challenge of long-sequence modeling. It achieves this by encoding each local block in parallel for computational speed while simultaneously and recurrently encoding global blocks to optimize GPU memory usage.
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+For the inference phase, the recurrent representation is used, favoring autoregressive decoding. This method efficiently reduces memory usage and latency, maintaining equivalent performance outcomes.
+
+## From Language to Image
+### RMT
+Paper [RMT: Retentive Networks Meet Vision Transformers](https://arxiv.org/abs/2309.11523) proposes a new vision backbone inspired by the RetNet architecture. The authors propose RMT to enhance the Vision Transformer (ViT) by introducing explicit spatial priors and reducing computational complexity, drawing inspiration from the RetNet's parallel representation. 
+This includes adapting the RetNet’s temporal decay to spatial domains and using a [Manhattan distance-based](https://en.wikipedia.org/wiki/Taxicab_geometry) spatial decay matrix, along with an attention decomposition form, to improve efficiency and scalability in vision tasks.
+
+* Manhattan Self-Attention (MaSA)
+![Attention Comparison.png](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/Attention%20Comparison.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:28]:
```suggestion
![Attention Comparison](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/Attention%20Comparison.png)
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+MaSA incorporates Self-Attention mechanism with a two-dimensional bidirectional spatial decay matrix based on the Manhattan distance among the tokens. This matrix decreases attention scores for tokens further away from a target token, allowing it to perceive global information while varying attention based on distance.
+
+* Decomposed Manhattan Self-Attention (MaSAD)
+![MaSAD.png](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/MaSAD.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:32]:
```suggestion
![MaSAD](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/MaSAD.png)
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+This mechanism decomposes Self-Attention in images along horizontal and vertical axes of the image, maintaining the spatial decay matrix without losing prior information. This decomposition allows the Manhattan Self-Attention (MaSA) to model global information efficiently with linear complexity, while preserving the original MaSA's receptive field shape.
+
+However, unlike the original RetNet that performs training with parallel representation and inference with recurrent representation, RMT does both processes with MaSA mechanism. The authors have done comparisons between MaSA and other RetNet's representations, and they show that MaSA has the best throughput with the highest accuracy.
+![MaSA vs Retention.png](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/MaSA%20vs%20Retention.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:36]:
```suggestion
![MaSA vs Retention](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/MaSA%20vs%20Retention.png)
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+### ViR
+![ViR.png](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/ViR.png)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:39]:
```suggestion
![ViR](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/ViR.png)
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+During the training phase, the approach incorporates both parallel and chunkwise recurrent representations, optimizing GPU usage for fast computation and being particularly effective for long sequences in terms of computational efficiency and memory use. 
+For the inference phase, the recurrent representation is used, favoring autoregressive decoding. This method efficiently reduces memory usage and latency, maintaining equivalent performance outcomes.
+
+## From Language to Image
+### RMT
+Paper [RMT: Retentive Networks Meet Vision Transformers](https://arxiv.org/abs/2309.11523) proposes a new vision backbone inspired by the RetNet architecture. The authors propose RMT to enhance the Vision Transformer (ViT) by introducing explicit spatial priors and reducing computational complexity, drawing inspiration from the RetNet's parallel representation. 
+This includes adapting the RetNet’s temporal decay to spatial domains and using a [Manhattan distance-based](https://en.wikipedia.org/wiki/Taxicab_geometry) spatial decay matrix, along with an attention decomposition form, to improve efficiency and scalability in vision tasks.
+
+* Manhattan Self-Attention (MaSA)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:27]:
```suggestion
- **Manhattan Self-Attention (MaSA)**
```

File: chapters/en/Unit 13 - Outlook/RMT.mdx [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+![Attention Comparison.png](https://huggingface.co/datasets/hf-vision/course-assets/blob/main/Attention%20Comparison.png)
+MaSA incorporates Self-Attention mechanism with a two-dimensional bidirectional spatial decay matrix based on the Manhattan distance among the tokens. This matrix decreases attention scores for tokens further away from a target token, allowing it to perceive global information while varying attention based on distance.
+
+* Decomposed Manhattan Self-Attention (MaSAD)
```
Comment (merveenoyan) [inline chapters/en/Unit 13 - Outlook/RMT.mdx:31]:
```suggestion
- **Decomposed Manhattan Self-Attention (MaSAD)**
```

**Review by merveenoyan — COMMENTED** [review]
Thank you! I mostly left formatting & readability nits

**Review by merveenoyan — APPROVED** [review]
LGTM! would you like to add a minimal implementation of a RetNet?

**Comment by farrosalferro** [comment]
Sure! But should I do it in this file or do it in a jupyter notebook file? I actually plan to create a jupyter notebook for implementing the code and doing comparison between RMT and ViR. Also thank you for the approval!

**Review by johko — APPROVED** [review]
LGTM :+1: 
Very crisp summary, thank you :hugs: 

Just one question: are there any git repos for the models you could link? I guess people would also liek to look at code

**Comment by farrosalferro** [comment]
Sure! I will add it later to the notebook that I mentioned in the previous comment! Thank you very much for the comment and approval!

**Comment by merveenoyan** [comment]
@farrosalferro I will merge this, you can add the minimal implementation to the mdx file, doesn't have to be anything extensive (you can checkout DETR section for a good example)

**Pull request merged**